### PR TITLE
Fix issue where closed TCP connections would remain in connection tracking table much longer than intended

### LINKED ISSUE #6280: Investigate TCP CT table cleanup for short-lived connections
### General Information

- Cilium version: 1.3.0
- Kernel version: 4.9 (kops debian stretch)
- Orchestration: k8s

### Summary

A user in Slack reported the TCP CT table becoming full, with these messages in Cilium v1.3.0:

```
level=warning msg="Garbage collection on ipv4 TCP CT map failed to finish" interrupted=0 subsys=map-ct
```
```
xx drop (CT: Map insertion failed) 
```

TCP-based DNS is in use, and due to the kernel version there is no LRU.

Even with TCP-based DNS, the connections should be cleaned up once the connection completes, however it appears that these entries are left in the CT table for a prolonged period.

Upon deeper inspection of the `cilium bpf ct list global` output, the vast majority (ALL TCP connections?) have `Flags=10`.

### PR DESCRIPTION
**Pull request by joestringer**
Previously, access into tcp flags was governed using a bitfield declared
based on the endianness of the host CPU, even though the packet data is
always in network byte-order. This would mean that any direct access of
flags would access the wrong bits in the packet, and would cause
conntrack entries for closed TCP connections to expire as long as 12
hours after a connection was closed.

Fix this issue by redefining the tcp flags struct to store in a 32-bit
structure, then use the Linux `TCP_FLAG_*` defines to check / store the
appropriate TCP flag bits.

We also ensure that the final ACK packet in a TCP termination will not
convert the connection back to a live connection (resetting the timer).

Individual commit messages have further detail, review by commit.

Fixes: #6280

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6318)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

File: bpf/lib/conntrack.h [hunk 1]
```diff
@@ -258,7 +253,7 @@ static inline int __inline__ __ct_lookup(void *map, struct __sk_buff *skb,
 		switch (action) {
 		case ACTION_CREATE:
 			ret = entry->rx_closing + entry->tx_closing;
-			if (unlikely(ret >= 1)) {
+			if (unlikely(ret >= 1 && (seen_flags.value & TCP_FLAG_SYN))) {
```
Comment (joestringer) [inline bpf/lib/conntrack.h]: I should call out this change explicitly.

A typical TCP connection close looks something like:
```
-> FIN
<- ACK, FIN
-> ACK
```
or
```
-> FIN
<- ACK
<- FIN
-> ACK
```

As a result, before the final ACK, the entry is likely to end up with `rx_closing` and `tx_closing` set; For the final ack packet, it does not have a `RST` or `FIN`, so the action ends up being `ACTION_CREATE` due to the following code:

https://github.com/cilium/cilium/blob/f0fdb07725f771be28f330519503f4818a251492/bpf/lib/conntrack.h#L377-L380

As a result, with the existing logic, the final ack will _always refresh the timer_ back to the full 12-hour TCP timeout, after the final ack previously triggered a timeout reduction in the `ACTION_CLOSE` case below.

This change alleviates this by only resetting the closing state and timeout if it appears that a brand new connection is establishing with the same 5-tuple.

Comment (joestringer) [reply bpf/lib/conntrack.h]: I put this explanation into a separate commit.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
@borkmann I believe that the definitions of the TCP_FLAG_XXX should handle CPU endianness issues, however I've only tested this on my own x86-64 machine.

**Comment by joestringer** [comment]
Seems like the extension of the `tcp_flags` to 32 bits increases the tail call complexity, causing the limit to be hit with IPv6 on kernel 4.9.

**Comment by joestringer** [comment]
test-me-please

EDIT:
```
Prog section '2/7' rejected: Invalid argument (22)!
 - Type:         3
 - Instructions: 4108 (12 over limit)
 - License:      GPL
```

I'll work on combining the checks for connection reopen in the final patch using bitwise operations.

**Comment by joestringer** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Comment by borkmann** [comment]
@joestringer Hmm, do you happen to know why exactly we ended up like this?

Similar issue as in: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=78a5a93c1eeb4e6933d1f62b33e5496d53b46c5a ?

**Comment by joestringer** [comment]
@borkmann I traced as far as `<asm/byte-order.h>` and `<linux/byteorder/little_endian.h>` which at least on my system unconditionally defines `__LITTLE_ENDIAN_BITFIELD`. Looking at the corresponding definitions for tcp flags in `struct tcphdr` in the Linux code, it seems like barely anything uses these bitfields, typically the Linux code instead takes a pointer to the word offset in the packet and uses `TCP_FLAG_XXX` to check flags, similar to what we're doing here. This leaves me pretty confused about how those linux uapi headers for TCP bitfields are supposed to work at all.

**Comment by joestringer** [comment]
Green checks all around, will rebase the `bpf.sha`.

**Review by borkmann — APPROVED** [review]

**Comment by borkmann** [comment]
> @borkmann I traced as far as `<asm/byte-order.h>` and `<linux/byteorder/little_endian.h>` which at least on my system unconditionally defines `__LITTLE_ENDIAN_BITFIELD`.

Interesting! So it does seem to compile in the assumed "correct" code presumably? On mu local setup adding a `#warning foo` I can see this shows up.

> Previously, access into tcp flags was governed using a bitfield declared based on the endianness of the host CPU, even though the packet data is always in network byte-order. This would mean that any direct access of flags would access the wrong bits in the packet.

I'm just confused a bit by the statement from the commit log. Were you dumping these bits and found that despite `__LITTLE_ENDIAN_BITFIELD` selection, their content didn't make any sense, e.g. is in the wrong order?

> Looking at the corresponding definitions for tcp flags in `struct tcphdr` in the Linux code, it seems like barely anything uses these bitfields, typically the Linux code instead takes a pointer to the word offset in the packet and uses `TCP_FLAG_XXX` to check flags, similar to what we're doing here. This leaves me pretty confused about how those linux uapi headers for TCP bitfields are supposed to work at all.

They normally should work as intended. Perhaps the issue here is that for native x86 host compiler and glibc these defines hold, but perhaps not necessarily for LLVM backend for BPF. Was this with old 3.8 LLVM?

**Pull request merged**